File: src/lib/battle/damage-calculator.ts
/**
 * Advanced Damage Calculator
 *
 * Calculates damage considering:
 * - Multiple damage types
 * - Resistances
 * - Accuracy/Evasion
 * - Critical hits
 * - Point defense interception
 * - Status effects
 */

import type {
  DamageTypes,
  StatusEffect,
  AdvancedDamageResult,
  DamageEvent,
  StatusEffectType,
  HackableSystem,
} from './damage-types'
import {
  EMPTY_DAMAGE,
  DAMAGE_EFFECTIVENESS,
  applyResistances,
  getTotalDamage,
  hasStatusEffect,
  getStatusEffectStrength,
} from './damage-types'
import type { AdvancedCombatUnit } from './advanced-unit'
import { canAttack, applyDamageToUnit } from './advanced-unit'

// ============================================================================
// CONSTANTS
// ============================================================================

/** Minimum hit chance (can't go below this) */
const MIN_HIT_CHANCE = 0.1

/** Minimum damage to penetrate shields (1% of shield) */
const MIN_SHIELD_PENETRATION = 0.01

/** Ionic damage threshold to trigger shield disruption */
const IONIC_DISRUPTION_THRESHOLD = 0.3

/** Hacking power threshold for system hack attempts */
const HACK_ATTEMPT_THRESHOLD = 50

/** Hull percentage threshold for hull breach */
const HULL_BREACH_THRESHOLD = 0.5

// ============================================================================
// MAIN DAMAGE CALCULATION
// ============================================================================

/**
 * Calculate advanced damage from attacker to defender
 */
export function calculateAdvancedDamage(
  attacker: AdvancedCombatUnit,
  defender: AdvancedCombatUnit,
  randomFn: () => number = Math.random
): AdvancedDamageResult {
  const events: DamageEvent[] = []
  const effects: StatusEffect[] = []

  // Check if attacker can attack
  if (!canAttack(attacker)) {
    return createMissResult()
  }

  // === POINT DEFENSE CHECK ===
  // Point defense can intercept small/fast units
  if (defender.stats.pointDefense > 0 && isInterceptable(attacker)) {
    const interceptChance = calculateInterceptChance(
      defender.stats.pointDefense,
      attacker.stats.evasion
    )
    if (randomFn() < interceptChance) {
      events.push({
        type: 'point_defense_intercept',
        data: { interceptedUnit: attacker.unitKey },
      })
      return {
        hit: false,
        critical: false,
        intercepted: true,
        rawDamage: { ...EMPTY_DAMAGE },
        finalDamage: { ...EMPTY_DAMAGE },
        appliedDamage: { shield: 0, armor: 0, hull: 0, crew: 0 },
        effects: [],
        events,
      }
    }
  }

  // === HIT CALCULATION ===
  const hitChance = calculateHitChance(attacker, defender)
  if (randomFn() > hitChance) {
    return createMissResult()
  }

  // === CRITICAL HIT CHECK ===
  const isCritical = randomFn() < attacker.stats.critChance / 100
  const critMultiplier = isCritical ? attacker.stats.critMultiplier : 1

  if (isCritical) {
    events.push({
      type: 'critical_hit',
      data: { multiplier: critMultiplier },
    })
  }

  // === RAW DAMAGE CALCULATION ===
  const rawDamage = calculateRawDamage(attacker, critMultiplier)

  // === APPLY RESISTANCES ===
  const finalDamage = applyResistances(rawDamage, defender.resistances)

  // === APPLY DAMAGE TO DEFENSE LAYERS ===
  const { shieldDamage, armorDamage, hullDamage, layerEvents } =
    calculateLayeredDamage(finalDamage, defender, events)

  events.push(...layerEvents)

  // Apply the damage
  const appliedDamage = applyDamageToUnit(
    defender,
    shieldDamage,
    armorDamage,
    hullDamage
  )

  // === STATUS EFFECTS ===

  // Ionic damage can disrupt shields
  if (finalDamage.ionic > 0) {
    const ionicEffect = checkIonicDisruption(
      finalDamage.ionic,
      defender,
      attacker.id
    )
    if (ionicEffect) {
      effects.push(ionicEffect)
    }
  }

  // Hacking attempts
  if (finalDamage.hacking >= HACK_ATTEMPT_THRESHOLD) {
    const hackResult = attemptHack(
      finalDamage.hacking,
      defender,
      attacker.id,
      randomFn
    )
    effects.push(...hackResult.effects)
    events.push(...hackResult.events)
  }

  // Hull breach check
  if (
    defender.defense.hull.current > 0 &&
    defender.defense.hull.current / defender.defense.hull.max < HULL_BREACH_THRESHOLD
  ) {
    const existingBreach = hasStatusEffect(defender.statusEffects, 'hull_breach')
    if (!existingBreach) {
      effects.push({
        type: 'hull_breach',
        duration: 3,
        strength: 0.2, // 20% increased damage taken
        sourceId: attacker.id,
      })
      events.push({
        type: 'hull_breach',
        data: { hullPercent: defender.defense.hull.current / defender.defense.hull.max },
      })
    }
  }

  // Crew casualties from hull damage
  let crewDamage = 0
  if (appliedDamage.hull > 0 && defender.crew.current > 0) {
    crewDamage = calculateCrewCasualties(
      appliedDamage.hull,
      defender.defense.hull.max,
      defender.crew.current,
      randomFn
    )
    if (crewDamage > 0) {
      defender.crew.current = Math.max(0, defender.crew.current - crewDamage)
      events.push({
        type: 'crew_casualties',
        data: { casualties: crewDamage, remaining: defender.crew.current },
      })
    }
  }

  // Apply new status effects to defender
  defender.statusEffects.push(...effects)

  return {
    hit: true,
    critical: isCritical,
    intercepted: false,
    rawDamage,
    finalDamage,
    appliedDamage: {
      shield: appliedDamage.shield,
      armor: appliedDamage.armor,
      hull: appliedDamage.hull,
      crew: crewDamage,
    },
    effects,
    events,
  }
}

// ============================================================================
// HIT CALCULATION
// ============================================================================

/**
 * Calculate hit chance based on accuracy and evasion
 */
function calculateHitChance(
  attacker: AdvancedCombatUnit,
  defender: AdvancedCombatUnit
): number {
  let accuracy = attacker.stats.accuracy
  let evasion = defender.stats.evasion

  // Apply status effect modifiers
  if (hasStatusEffect(attacker.statusEffects, 'ionized')) {
    accuracy *= 1 - getStatusEffectStrength(attacker.statusEffects, 'ionized')
  }

  // Base formula: (accuracy - evasion + 50) / 100
  const hitChance = (accuracy - evasion + 50) / 100

  return Math.max(MIN_HIT_CHANCE, Math.min(1, hitChance))
}

// ============================================================================
// RAW DAMAGE CALCULATION
// ============================================================================

/**
 * Calculate raw damage output from attacker
 */
function calculateRawDamage(
  attacker: AdvancedCombatUnit,
  critMultiplier: number
): DamageTypes {
  let damage = { ...attacker.damage }

  // Apply crit multiplier to direct damage types
  damage.ballistic = Math.floor(damage.ballistic * critMultiplier)
  damage.ionic = Math.floor(damage.ionic * critMultiplier)
  damage.explosive = Math.floor(damage.explosive * critMultiplier)
  // Hacking and boarding are not affected by crits

  // Apply status effect modifiers
  if (hasStatusEffect(attacker.statusEffects, 'ionized')) {
    const reduction = getStatusEffectStrength(attacker.statusEffects, 'ionized')
    damage.ballistic = Math.floor(damage.ballistic * (1 - reduction))
    damage.ionic = Math.floor(damage.ionic * (1 - reduction))
    damage.explosive = Math.floor(damage.explosive * (1 - reduction))
  }

  return damage
}

// ============================================================================
// LAYERED DAMAGE CALCULATION
// ============================================================================

interface LayeredDamageResult {
  shieldDamage: number
  armorDamage: number
  hullDamage: number
  layerEvents: DamageEvent[]
}

/**
 * Calculate how damage is distributed across defense layers
 */
function calculateLayeredDamage(
  damage: DamageTypes,
  defender: AdvancedCombatUnit,
  existingEvents: DamageEvent[]
): LayeredDamageResult {
  const events: DamageEvent[] = []

  // Calculate effective damage against each layer
  const vsShield =
    damage.ballistic * DAMAGE_EFFECTIVENESS.ballistic.vsShield +
    damage.ionic * DAMAGE_EFFECTIVENESS.ionic.vsShield +
    damage.explosive * DAMAGE_EFFECTIVENESS.explosive.vsShield

  const vsArmor =
    damage.ballistic * DAMAGE_EFFECTIVENESS.ballistic.vsArmor +
    damage.ionic * DAMAGE_EFFECTIVENESS.ionic.vsArmor +
    damage.explosive * DAMAGE_EFFECTIVENESS.explosive.vsArmor

  const vsHull =
    damage.ballistic * DAMAGE_EFFECTIVENESS.ballistic.vsHull +
    damage.ionic * DAMAGE_EFFECTIVENESS.ionic.vsHull +
    damage.explosive * DAMAGE_EFFECTIVENESS.explosive.vsHull

  // Check shield penetration
  const totalDamage = getTotalDamage(damage)
  const shieldCurrent = defender.defense.shield.current

  if (shieldCurrent > 0 && totalDamage < shieldCurrent * MIN_SHIELD_PENETRATION) {
    // Damage bounces off shields
    return {
      shieldDamage: 0,
      armorDamage: 0,
      hullDamage: 0,
      layerEvents: events,
    }
  }

  // Calculate damage flow through layers
  let shieldDamage = Math.min(shieldCurrent, vsShield)
  let remainingDamage = vsShield - shieldDamage

  // Check for shield break
  if (shieldCurrent > 0 && defender.defense.shield.current - shieldDamage <= 0) {
    events.push({
      type: 'shield_break',
      data: { shieldDamage },
    })
  }

  // Armor absorbs remaining shield overflow + direct armor damage
  const armorCurrent = defender.defense.armor.current
  let armorDamage = Math.min(armorCurrent, vsArmor + remainingDamage * 0.5)
  remainingDamage = Math.max(0, vsArmor + remainingDamage * 0.5 - armorCurrent)

  // Check for armor break
  if (armorCurrent > 0 && defender.defense.armor.current - armorDamage <= 0) {
    events.push({
      type: 'armor_break',
      data: { armorDamage },
    })
  }

  // Hull takes remaining damage
  const hullDamage = vsHull + remainingDamage

  // Apply hull breach modifier if present
  let finalHullDamage = hullDamage
  if (hasStatusEffect(defender.statusEffects, 'hull_breach')) {
    const breachStrength = getStatusEffectStrength(
      defender.statusEffects,
      'hull_breach'
    )
    finalHullDamage = Math.floor(hullDamage * (1 + breachStrength))
  }

  return {
    shieldDamage: Math.floor(shieldDamage),
    armorDamage: Math.floor(armorDamage),
    hullDamage: Math.floor(finalHullDamage),
    layerEvents: events,
  }
}

// ============================================================================
// IONIC DISRUPTION
// ============================================================================

/**
 * Check if ionic damage causes shield disruption
 */
function checkIonicDisruption(
  ionicDamage: number,
  defender: AdvancedCombatUnit,
  sourceId: string
): StatusEffect | null {
  if (defender.defense.shield.max === 0) return null

  const ionicRatio = ionicDamage / defender.defense.shield.max

  if (ionicRatio >= IONIC_DISRUPTION_THRESHOLD) {
    return {
      type: 'shield_disruption',
      duration: 1,
      strength: Math.min(0.5, ionicRatio), // Up to 50% shield regen reduction
      sourceId,
    }
  }

  return null
}

// ============================================================================
// HACKING
// ============================================================================

interface HackResult {
  effects: StatusEffect[]
  events: DamageEvent[]
}

/**
 * Attempt to hack enemy systems
 */
function attemptHack(
  hackPower: number,
  defender: AdvancedCombatUnit,
  sourceId: string,
  randomFn: () => number
): HackResult {
  const effects: StatusEffect[] = []
  const events: DamageEvent[] = []

  // Calculate hack success chance
  const hackDefense = defender.resistances.hack_defense
  const successChance = Math.max(0.1, (hackPower - hackDefense) / 100)

  if (randomFn() < successChance) {
    // Hack succeeded - select random system
    const targetSystem = selectRandomSystem(randomFn)
    const effect = createHackEffect(targetSystem, hackPower, sourceId)

    effects.push(effect)
    events.push({
      type: 'hack_success',
      data: { targetSystem, hackPower },
    })

    // Severe hacks can disable the unit
    if (hackPower > 100 && targetSystem === 'power_core') {
      defender.disabled = true
    }
  } else {
    events.push({
      type: 'hack_failed',
      data: { hackPower, hackDefense },
    })
  }

  return { effects, events }
}

/**
 * Select a random hackable system
 */
function selectRandomSystem(randomFn: () => number): HackableSystem {
  const systems: HackableSystem[] = [
    'weapons',
    'shields',
    'engines',
    'sensors',
    'communications',
  ]
  const index = Math.floor(randomFn() * systems.length)
  return systems[index]
}

/**
 * Create a status effect for a hacked system
 */
function createHackEffect(
  system: HackableSystem,
  hackPower: number,
  sourceId: string
): StatusEffect {
  const duration = Math.min(3, Math.floor(hackPower / 50))
  const strength = Math.min(1, hackPower / 100)

  const typeMap: Record<HackableSystem, StatusEffectType> = {
    weapons: 'weapons_disabled',
    shields: 'shield_disruption',
    engines: 'engines_disabled',
    sensors: 'ionized',
    communications: 'system_hacked',
    life_support: 'crew_panic',
    power_core: 'emp_stunned',
  }

  return {
    type: typeMap[system],
    duration,
    strength,
    sourceId,
    targetSystem: system,
  }
}

// ============================================================================
// POINT DEFENSE
// ============================================================================

/**
 * Check if unit can be intercepted by point defense
 */
function isInterceptable(unit: AdvancedCombatUnit): boolean {
  // Fighters and missiles can be intercepted
  return (
    unit.unitClass === 'fighter' ||
    unit.unitClass === 'missile' ||
    unit.unitKey === 'espionage_probe'
  )
}

/**
 * Calculate point defense intercept chance
 */
function calculateInterceptChance(
  pointDefense: number,
  targetEvasion: number
): number {
  // Base intercept chance from point defense value
  const baseChance = pointDefense / 100

  // Reduced by target evasion
  const adjustedChance = baseChance * (1 - targetEvasion / 200)

  return Math.max(0, Math.min(0.8, adjustedChance))
}

// ============================================================================
// CREW CASUALTIES
// ============================================================================

/**
 * Calculate crew casualties from hull damage
 */
function calculateCrewCasualties(
  hullDamage: number,
  maxHull: number,
  currentCrew: number,
  randomFn: () => number
): number {
  // Crew casualties proportional to hull damage
  const damageRatio = hullDamage / maxHull
  const baseCasualties = Math.floor(currentCrew * damageRatio * 0.5)

  // Add some randomness
  const variance = Math.floor(randomFn() * baseCasualties * 0.2)
  const casualties = baseCasualties + variance - Math.floor(baseCasualties * 0.1)

  return Math.max(0, Math.min(currentCrew, casualties))
}

// ============================================================================
// HELPERS
// ============================================================================

/**
 * Create a miss result
 */
function createMissResult(): AdvancedDamageResult {
  return {
    hit: false,
    critical: false,
    intercepted: false,
    rawDamage: { ...EMPTY_DAMAGE },
    finalDamage: { ...EMPTY_DAMAGE },
    appliedDamage: { shield: 0, armor: 0, hull: 0, crew: 0 },
    effects: [],
    events: [],
  }
}

// ============================================================================
// LEGACY COMPATIBILITY
// ============================================================================

/**
 * Convert legacy damage result to simple values for backward compatibility
 */
export function simplifyDamageResult(result: AdvancedDamageResult): {
  shieldDamage: number
  hullDamage: number
  bounced: boolean
} {
  return {
    shieldDamage: result.appliedDamage.shield,
    hullDamage: result.appliedDamage.armor + result.appliedDamage.hull,
    bounced: !result.hit,
  }
}

/**
 * Calculate simple damage (legacy compatibility)
 * Converts total weapon power to damage types based on unit class
 */
export function convertLegacyDamage(
  weaponPower: number,
  unitClass: string
): DamageTypes {
  // Default damage distribution based on unit type
  const distributions: Record<string, { b: number; i: number; e: number }> = {
    fighter: { b: 0.8, i: 0.1, e: 0.1 },
    corvette: { b: 0.7, i: 0.2, e: 0.1 },
    frigate: { b: 0.6, i: 0.2, e: 0.2 },
    cruiser: { b: 0.5, i: 0.3, e: 0.2 },
    battleship: { b: 0.4, i: 0.3, e: 0.3 },
    carrier: { b: 0.3, i: 0.4, e: 0.3 },
    dreadnought: { b: 0.3, i: 0.3, e: 0.4 },
    defense: { b: 0.5, i: 0.3, e: 0.2 },
    missile: { b: 0.1, i: 0.0, e: 0.9 },
    transport: { b: 1.0, i: 0.0, e: 0.0 },
    utility: { b: 1.0, i: 0.0, e: 0.0 },
  }

  const dist = distributions[unitClass] ?? distributions.fighter

  return {
    ballistic: Math.floor(weaponPower * dist.b),
    ionic: Math.floor(weaponPower * dist.i),
    explosive: Math.floor(weaponPower * dist.e),
    hacking: 0,
    boarding: 0,
  }
}

File: unity_client/Assets/Scripts/SharedLib/Battle/DamageCalculator.cs
// =============================================================================
// DamageCalculator.cs — Ported from OGameX src/lib/battle/damage-calculator.ts
// Advanced damage calculator: accuracy, crits, layered damage, hacking, crew
// =============================================================================

using System;
using System.Collections.Generic;
using System.Linq;

namespace OGameX.SharedLib.Battle
{
    /// <summary>
    /// Advanced damage calculator with multi-type damage, accuracy/evasion,
    /// critical hits, point defense interception, status effects (ionic
    /// disruption, hacking, hull breach), and crew casualties.
    /// <para>Ported from src/lib/battle/damage-calculator.ts.</para>
    /// </summary>
    public static class DamageCalculator
    {
        // Constants
        private const double MinHitChance = 0.10;
        private const double MinShieldPenetration = 0.01;
        private const double IonicDisruptionThreshold = 0.30;
        private const int HackAttemptThreshold = 50;
        private const double HullBreachThreshold = 0.50;

        // =====================================================================
        // MAIN ENTRY POINT
        // =====================================================================

        /// <summary>
        /// Calculate advanced damage from attacker to defender.
        /// </summary>
        public static AdvancedDamageResult CalculateAdvancedDamage(
            AdvancedCombatUnit attacker, AdvancedCombatUnit defender, Random rng)
        {
            var events = new List<DamageEvent>();
            var effects = new List<StatusEffect>();

            if (!AdvancedUnitHelpers.CanAttack(attacker))
                return MissResult();

            // --- Point defense interception ---
            if (defender.Stats.PointDefense > 0 && IsInterceptable(attacker))
            {
                double interceptChance = CalculateInterceptChance(
                    defender.Stats.PointDefense, attacker.Stats.Evasion);
                if (rng.NextDouble() < interceptChance)
                {
                    events.Add(new DamageEvent
                    {
                        Type = DamageEventType.PointDefenseIntercept,
                        Data = { ["interceptedUnit"] = attacker.UnitKey },
                    });
                    return new AdvancedDamageResult
                    {
                        Hit = false, Intercepted = true, Events = events,
                    };
                }
            }

            // --- Hit check ---
            double hitChance = CalculateHitChance(attacker, defender);
            if (rng.NextDouble() > hitChance)
                return MissResult();

            // --- Critical hit ---
            bool isCrit = rng.NextDouble() < attacker.Stats.CritChance / 100.0;
            double critMult = isCrit ? attacker.Stats.CritMultiplier : 1.0;
            if (isCrit)
                events.Add(new DamageEvent
                {
                    Type = DamageEventType.CriticalHit,
                    Data = { ["multiplier"] = critMult },
                });

            // --- Raw damage ---
            var rawDamage = CalculateRawDamage(attacker, critMult);

            // --- Resistances ---
            var finalDamage = DamageTypeHelpers.ApplyResistances(rawDamage, defender.Resistances);

            // --- Layered damage (shield → armor → hull) ---
            var (shieldDmg, armorDmg, hullDmg, layerEvents) =
                CalculateLayeredDamage(finalDamage, defender);
            events.AddRange(layerEvents);

            // Apply damage
            var (actShield, actArmor, actHull) =
                AdvancedUnitHelpers.ApplyDamage(defender, shieldDmg, armorDmg, hullDmg);

            // --- Status effects ---
            // Ionic disruption
            if (finalDamage.Ionic > 0)
            {
                var ionEffect = CheckIonicDisruption(finalDamage.Ionic, defender, attacker.Id);
                if (ionEffect != null) effects.Add(ionEffect);
            }

            // Hacking
            if (finalDamage.Hacking >= HackAttemptThreshold)
            {
                var (hackEffects, hackEvents) = AttemptHack(
                    finalDamage.Hacking, defender, attacker.Id, rng);
                effects.AddRange(hackEffects);
                events.AddRange(hackEvents);
            }

            // Hull breach
            if (defender.Defense.Hull.Current > 0 &&
                (double)defender.Defense.Hull.Current / defender.Defense.Hull.Max < HullBreachThreshold &&
                !DamageTypeHelpers.HasEffect(defender.StatusEffects, StatusEffectType.HullBreach))
            {
                effects.Add(new StatusEffect
                {
                    Type = StatusEffectType.HullBreach,
                    Duration = 3,
                    Strength = 0.2,
                    SourceId = attacker.Id,
                });
                events.Add(new DamageEvent
                {
                    Type = DamageEventType.HullBreach,
                    Data = { ["hullPercent"] = (double)defender.Defense.Hull.Current / defender.Defense.Hull.Max },
                });
            }

            // Crew casualties
            int crewDmg = 0;
            if (actHull > 0 && defender.CrewCurrent > 0)
            {
                crewDmg = CalculateCrewCasualties(actHull, defender.Defense.Hull.Max, defender.CrewCurrent, rng);
                if (crewDmg > 0)
                {
                    defender.CrewCurrent = Math.Max(0, defender.CrewCurrent - crewDmg);
                    events.Add(new DamageEvent
                    {
                        Type = DamageEventType.CrewCasualties,
                        Data = { ["casualties"] = crewDmg, ["remaining"] = defender.CrewCurrent },
                    });
                }
            }

            // Push status effects onto defender
            defender.StatusEffects.AddRange(effects);

            return new AdvancedDamageResult
            {
                Hit = true,
                Critical = isCrit,
                Intercepted = false,
                RawDamage = rawDamage,
                FinalDamage = finalDamage,
                AppliedShield = actShield,
                AppliedArmor = actArmor,
                AppliedHull = actHull,
                AppliedCrew = crewDmg,
                Effects = effects,
                Events = events,
            };
        }

        // =====================================================================
        // HIT CALCULATION
        // =====================================================================

        private static double CalculateHitChance(AdvancedCombatUnit attacker, AdvancedCombatUnit defender)
        {
            double accuracy = attacker.Stats.Accuracy;
            if (DamageTypeHelpers.HasEffect(attacker.StatusEffects, StatusEffectType.Ionized))
                accuracy *= 1.0 - DamageTypeHelpers.GetEffectStrength(attacker.StatusEffects, StatusEffectType.Ionized);

            double hit = (accuracy - defender.Stats.Evasion + 50.0) / 100.0;
            return Math.Max(MinHitChance, Math.Min(1.0, hit));
        }

        // =====================================================================
        // RAW DAMAGE
        // =====================================================================

        private static DamageTypes CalculateRawDamage(AdvancedCombatUnit attacker, double critMult)
        {
            var d = attacker.Damage;
            int b = (int)Math.Floor(d.Ballistic * critMult);
            int i = (int)Math.Floor(d.Ionic * critMult);
            int e = (int)Math.Floor(d.Explosive * critMult);

            if (DamageTypeHelpers.HasEffect(attacker.StatusEffects, StatusEffectType.Ionized))
            {
                double red = DamageTypeHelpers.GetEffectStrength(attacker.StatusEffects, StatusEffectType.Ionized);
                b = (int)Math.Floor(b * (1.0 - red));
                i = (int)Math.Floor(i * (1.0 - red));
                e = (int)Math.Floor(e * (1.0 - red));
            }

            return new DamageTypes
            {
                Ballistic = b,
                Ionic = i,
                Explosive = e,
                Hacking = d.Hacking,
                Boarding = d.Boarding,
            };
        }

        // =====================================================================
        // LAYERED DAMAGE
        // =====================================================================

        private static (int ShieldDmg, int ArmorDmg, int HullDmg, List<DamageEvent> Events)
            CalculateLayeredDamage(DamageTypes dmg, AdvancedCombatUnit defender)
        {
            var events = new List<DamageEvent>();

            double vsShield = dmg.Ballistic * DamageEffectiveness.BallisticVsShield
                            + dmg.Ionic * DamageEffectiveness.IonicVsShield
                            + dmg.Explosive * DamageEffectiveness.ExplosiveVsShield;
            double vsArmor = dmg.Ballistic * DamageEffectiveness.BallisticVsArmor
                           + dmg.Ionic * DamageEffectiveness.IonicVsArmor
                           + dmg.Explosive * DamageEffectiveness.ExplosiveVsArmor;
            double vsHull = dmg.Ballistic * DamageEffectiveness.BallisticVsHull
                          + dmg.Ionic * DamageEffectiveness.IonicVsHull
                          + dmg.Explosive * DamageEffectiveness.ExplosiveVsHull;

            int shieldCur = defender.Defense.Shield.Current;
            int totalDmg = dmg.TotalCombatDamage;

            // Bounce
            if (shieldCur > 0 && totalDmg < shieldCur * MinShieldPenetration)
                return (0, 0, 0, events);

            int shieldDamage = Math.Min(shieldCur, (int)vsShield);
            double overflow = vsShield - shieldDamage;

            if (shieldCur > 0 && defender.Defense.Shield.Current - shieldDamage <= 0)
                events.Add(new DamageEvent { Type = DamageEventType.ShieldBreak });

            int armorCur = defender.Defense.Armor.Current;
            int armorDamage = Math.Min(armorCur, (int)(vsArmor + overflow * 0.5));
            double armorOverflow = Math.Max(0, vsArmor + overflow * 0.5 - armorCur);

            if (armorCur > 0 && defender.Defense.Armor.Current - armorDamage <= 0)
                events.Add(new DamageEvent { Type = DamageEventType.ArmorBreak });

            double rawHull = vsHull + armorOverflow;
            if (DamageTypeHelpers.HasEffect(defender.StatusEffects, StatusEffectType.HullBreach))
            {
                double str = DamageTypeHelpers.GetEffectStrength(defender.StatusEffects, StatusEffectType.HullBreach);
                rawHull = Math.Floor(rawHull * (1.0 + str));
            }

            return (shieldDamage, armorDamage, (int)rawHull, events);
        }

        // =====================================================================
        // IONIC DISRUPTION
        // =====================================================================

        private static StatusEffect CheckIonicDisruption(int ionicDmg, AdvancedCombatUnit defender, string srcId)
        {
            if (defender.Defense.Shield.Max == 0) return null;
            double ratio = (double)ionicDmg / defender.Defense.Shield.Max;
            if (ratio < IonicDisruptionThreshold) return null;
            return new StatusEffect
            {
                Type = StatusEffectType.ShieldDisruption,
                Duration = 1,
                Strength = Math.Min(0.5, ratio),
                SourceId = srcId,
            };
        }

        // =====================================================================
        // HACKING
        // =====================================================================

        private static readonly HackableSystem[] HackableSystems =
        {
            HackableSystem.Weapons, HackableSystem.Shields, HackableSystem.Engines,
            HackableSystem.Sensors, HackableSystem.Communications,
        };

        private static readonly Dictionary<HackableSystem, StatusEffectType> HackEffectMap =
            new Dictionary<HackableSystem, StatusEffectType>
        {
            { HackableSystem.Weapons, StatusEffectType.WeaponsDisabled },
            { HackableSystem.Shields, StatusEffectType.ShieldDisruption },
            { HackableSystem.Engines, StatusEffectType.EnginesDisabled },
            { HackableSystem.Sensors, StatusEffectType.Ionized },
            { HackableSystem.Communications, StatusEffectType.SystemHacked },
            { HackableSystem.LifeSupport, StatusEffectType.CrewPanic },
            { HackableSystem.PowerCore, StatusEffectType.EmpStunned },
        };

        private static (List<StatusEffect>, List<DamageEvent>) AttemptHack(
            int hackPower, AdvancedCombatUnit defender, string srcId, Random rng)
        {
            var effects = new List<StatusEffect>();
            var events = new List<DamageEvent>();

            double chance = Math.Max(0.1, (hackPower - defender.Resistances.HackDefense) / 100.0);

            if (rng.NextDouble() < chance)
            {
                var sys = HackableSystems[rng.Next(HackableSystems.Length)];
                int dur = Math.Min(3, (int)Math.Floor((double)hackPower / 50));
                double str = Math.Min(1.0, hackPower / 100.0);

                effects.Add(new StatusEffect
                {
                    Type = HackEffectMap[sys],
                    Duration = dur,
                    Strength = str,
                    SourceId = srcId,
                    TargetSystem = sys,
                });
                events.Add(new DamageEvent
                {
                    Type = DamageEventType.HackSuccess,
                    Data = { ["targetSystem"] = sys.ToString(), ["hackPower"] = hackPower },
                });

                if (hackPower > 100 && sys == HackableSystem.PowerCore)
                    defender.Disabled = true;
            }
            else
            {
                events.Add(new DamageEvent
                {
                    Type = DamageEventType.HackFailed,
                    Data = { ["hackPower"] = hackPower, ["hackDefense"] = defender.Resistances.HackDefense },
                });
            }
            return (effects, events);
        }

        // =====================================================================
        // POINT DEFENSE
        // =====================================================================

        private static bool IsInterceptable(AdvancedCombatUnit u) =>
            u.UnitClass == UnitClass.Fighter ||
            u.UnitClass == UnitClass.Missile ||
            u.UnitKey == "espionage_probe";

        private static double CalculateInterceptChance(int pointDefense, double targetEvasion)
        {
            double baseChance = pointDefense / 100.0;
            double adj = baseChance * (1.0 - targetEvasion / 200.0);
            return Math.Max(0, Math.Min(0.8, adj));
        }

        // =====================================================================
        // CREW CASUALTIES
        // =====================================================================

        private static int CalculateCrewCasualties(int hullDmg, int maxHull, int currentCrew, Random rng)
        {
            double damageRatio = (double)hullDmg / maxHull;
            int baseCas = (int)Math.Floor(currentCrew * damageRatio * 0.5);
            int variance = (int)Math.Floor(rng.NextDouble() * baseCas * 0.2);
            int total = baseCas + variance - (int)Math.Floor(baseCas * 0.1);
            return Math.Max(0, Math.Min(currentCrew, total));
        }

        // =====================================================================
        // LEGACY COMPATIBILITY
        // =====================================================================

        /// <summary>
        /// Convert total weapon power into 5-channel DamageTypes based on unit class.
        /// </summary>
        public static DamageTypes ConvertLegacyDamage(int weaponPower, UnitClass unitClass)
        {
            (double b, double i, double e) dist = unitClass switch
            {
                UnitClass.Fighter => (0.8, 0.1, 0.1),
                UnitClass.Corvette => (0.7, 0.2, 0.1),
                UnitClass.Frigate => (0.6, 0.2, 0.2),
                UnitClass.Cruiser => (0.5, 0.3, 0.2),
                UnitClass.Battleship => (0.4, 0.3, 0.3),
                UnitClass.Carrier => (0.3, 0.4, 0.3),
                UnitClass.Dreadnought => (0.3, 0.3, 0.4),
                UnitClass.Defense => (0.5, 0.3, 0.2),
                UnitClass.Missile => (0.1, 0.0, 0.9),
                UnitClass.Transport => (1.0, 0.0, 0.0),
                _ => (1.0, 0.0, 0.0),
            };

            return new DamageTypes
            {
                Ballistic = (int)Math.Floor(weaponPower * dist.b),
                Ionic = (int)Math.Floor(weaponPower * dist.i),
                Explosive = (int)Math.Floor(weaponPower * dist.e),
                Hacking = 0,
                Boarding = 0,
            };
        }

        // =====================================================================
        // MISS RESULT
        // =====================================================================

        private static AdvancedDamageResult MissResult() => new AdvancedDamageResult
        {
            Hit = false, Critical = false, Intercepted = false,
        };
    }
}
